resources/type-system/inference.md

```diff
@@ -0,0 +1,830 @@
+(declared or inferred) for a parameter position which is being inferred for the
+overriding method, it is an error.  If there is no corresponding parameter
+position in the overridden method to infer from and the signatures are
+compatible, it is treated as dynamic (e.g. overriding a one parameter method
+with a method that takes a second optional parameter).  Note: if there is no
+corresponding parameter position in the overriden method to infer from and the
+signatures are incompatible (e.g. overriding a one parameter method with a
+method that takes a second non-optional parameter), the inference result is not
+defined and tools are free to either emit an error, or to defer the error to
+override checking.
+
+
+##### Instance field, getter, and setter override inference
+
+The inferred type of a getter, setter, or field is computed as follows.  Note
+that we say that a setter overrides a getter if there is a getter of the same
+name in some superclass or interface (explicitly declared or induced by an
+instance variable declaration), and similarly for setters overriding getters,
+fields, etc.
+
+A getter, setter or field which overrides/implements only a getter is inferred
+to have the type taken from the overridden getter result type.
+
+A getter, setter or field which overrides/implements only a setter is inferred
+to have the type taken from the overridden setter parameter.
+
+A getter which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden getter result type.
+
+A setter which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden setter parameter type.
+
+A final field which overrides/implements both a setter and a getter is inferred
+to have the type taken from the overridden getter result type.
+
+A non-final field which overrides/implements both a setter and a getter is
+inferred to have the type taken from the overridden setter parameter type if
+this type is the same as the return type of the overridden getter (if the types
+are not the same then inference fails with an error).
+
+Note that overriding a field is addressed via the implicit induced getter/setter
+pair (or just getter in the case of a final field).
+
+## Local type inference
+
+When types are omitted on local variable declarations and function expressions,
+or when type arguments are omitted from literal expressions, constructor
+invocations, or generic function invocations, then local type inference is used
+to fill in the missing type information.  Local type inference is also used as
+part of top-level inference as described above, in order to infer types for
+initializer expressions.  In order to uniformly treat use of local type
+inference in top-level inference and in method body inference, it is defined
+with respect to a set of *available* variables as defined above.  Note however
+that top-level inference never depends on method body inference, and so method
```

Comment:
Just noting that this property seems to be violated by the current implementation (`dartanalyzer`, commit 9a4c49f30118e7a53728acb4479cf30e29c2087b):

```dart
X id<X>(X x) => x;
class A { A(); }
var x = () => id(y);
var y = new A();

main() {
  x = 42; // Error, ".. can't be assigned to .. '() → A'"
}
```
In order to report the error in `main`, inference must have given `x` the type `A Function()`, which would hardly happen unless the body of the function literal were consulted in order to compute the return type, which will in turn require the invocation of `id` to receive inferred type arguments, and, of course, relies on the type of `y` which is also inferred.

Does that not count?

Reply:
No, I mean methods specifically as, well, methods (that is, instance members of classes which are not fields, getters, or setters).  x is a field which is initialized with a function literal, and inference does depend on function literal inference.